### PR TITLE
Adding door connections

### LINKED ISSUE #67: Door connections and locked doors
## Expected Behaviour
Allow Sherry to move between rooms.
Add doors to rooms to indicate how Sherry can leave/enter a room.
Some doors will be locked for level 1.
When Sherry goes between rooms the game scene will go black then reappear in the new room.

## Actual Behaviour
Currently, Sherry is only in one room and there are no doors in rooms.

## Steps to Reproduce Behaviour
1. Start the game, Sherry is only in the Hall and there are no doors in any room.

### PR DESCRIPTION
**Pull request by Alice-Jaffray**
**Description of work.**

Allowing Sherry to walk through rooms using the doors.
The door connections follow this map:

WhatsApp Image 2021-01-17 at 12 43 03

**To test:**

Start the game and get Sherry to walk through the doors and into another room

Fixes #67

### CODE REVIEW

File: project/sherLOCKED/Assets/Scripts/Objects/DoorInteraction.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
+using UnityEngine.SceneManagement;
+using UnityEngine;
+using UnityEngine.UI;
+
+public class DoorInteraction : MonoBehaviour
+{
+    private bool inCollider = false;
+
+    void Start() {
+    }
```
Comment (cailafinn) [inline project/sherLOCKED/Assets/Scripts/Objects/DoorInteraction.cs:15]: Can be deleted if body is empty

File: project/sherLOCKED/Assets/Scripts/Objects/DoorInteraction.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    // checks if player leaves the door perimeter
+    // if player does leave the perimeter the set collider variable to false
+    void OnTriggerExit2D(Collider2D other)
+    {
+        if (other.gameObject.CompareTag("Player"))
+        {
+            inCollider = false;
+        }
+    }
+}
```
Comment (cailafinn) [inline project/sherLOCKED/Assets/Scripts/Objects/DoorInteraction.cs]: Should have one blank line at the end of the file

**Review by cailafinn — CHANGES_REQUESTED** [review]
It's a really good start, but there are some notes:
- The door assets have not had the compression and filter settings set according to the asset plan (No compression, and point)
- Sherry appears on the wall, rather than the floor in front of the door after going through it
- Sherry cannot return through the door

Further Question: Do we want to go through doors automatically, or do we want to press E or Space to interact with the door to walk through it? I think the latter would be preferable but I'm unsure.

**Review by cailafinn — CHANGES_REQUESTED** [review]
Movement between rooms looks good. Just a couple more notes:
- There's currently a merge conflict with master, which needs resolving
- Sherry returns slightly too far away from the door when they return to the hall. 
- Sherry's duplication issue may be solved by this: https://answers.unity.com/questions/982403/how-to-not-duplicate-game-objects-on-dontdestroyon.html

File: project/sherLOCKED/Assets/Scripts/HallStart.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        //sherryInteract.interact = interactionSymbol;
+        // Adding a BoxCollider2D used by composite
+        BoxCollider2D bc2 = sherryPrefab.AddComponent<BoxCollider2D>();
+        bc2.offset = new Vector2(0, -0.5f);
+        bc2.size = new Vector2(1, 0.25f);
+        //bc2.usedByComposite = true;
+        // Adding an animator to the player
+        //sherry.AddComponent<Animator>();
+        Animator anim = sherryPrefab.AddComponent<Animator>();
+        var controller = UnityEditor.Animations.AnimatorController.CreateAnimatorControllerAtPath("Assets/Animations/sherry_ac.controller");
```
Comment (cailafinn) [inline project/sherLOCKED/Assets/Scripts/HallStart.cs]: **!!! This line is DELETING the animation controller, replacing it with a blank one!**

This is why sherry is no longer animating
The animation controller needs to be reset to the one currently on master (it's git ref is `f1324d9`)

Comment (cailafinn) [inline project/sherLOCKED/Assets/Scripts/HallStart.cs]: All of this should be set in the prefab, not the script (which has been done on master). 

Only three things need setting  in the script that cannot be set in the prefab:
- The Reputation bar object in the scene
- The Intuition bar object in the scene
- The interaction symbol UI element

File: project/sherLOCKED/Assets/Scripts/HallStart.cs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+using UnityEditor.Animations;
+
+public class HallStart : MonoBehaviour
+{
+    // Reference to the Sherry prefab
+    public GameObject sherryPrefab;
+    // public float walkSpeed;
+    // public GameObject interact;
+
+    void Start()
```
Comment (cailafinn) [inline project/sherLOCKED/Assets/Scripts/HallStart.cs:24]: This method only really needs to contain these lines:
```
sherry = Instantiate(sherryPrefab, new Vector3(0, 0, 0), Quaternion.identity);
sherry.GetComponent<SherryInteract>().interact = GameObject.Find("interactionSymbol");
sherry.GetComponent<SherryProgress>().intuitionBar = GameObject.Find("IntuitionBar").GetComponent<ProgressBar>();  
sherry.GetComponent<SherryProgress>().reputationBar = GameObject.Find("ReputationBar").GetComponent<ProgressBar>();
```

**Review by cailafinn — CHANGES_REQUESTED** [review]
Some experimentation has also revealed that the computer interation script also needs a line and variable (`prog`) to find the player object's progress tracking script, as it can't be assigned manually:
`prog = other.gameObject.GetComponent<SherryProgress>();` in `OnTriggerEnter2D`

File: project/sherLOCKED/Assets/Scripts/Objects/DoorInteraction.cs [hunk 5]
```diff
@@ -20,22 +24,90 @@ void OnTriggerEnter2D(Collider2D other)
         Scene currentRoom = SceneManager.GetActiveScene ();
         // Get the name of this room
         string roomName = currentRoom.name;
+
+        // // Get the name of the door
+        // string doorName =
```
Comment (cailafinn) [inline project/sherLOCKED/Assets/Scripts/Objects/DoorInteraction.cs]: Can some of this commented out code be removed?

File: project/sherLOCKED/Assets/Scripts/Objects/DoorInteraction.cs [hunk 6]
```diff
@@ -20,22 +24,90 @@ void OnTriggerEnter2D(Collider2D other)
...
-            playerPos = transform.position;
-            playerPos.x = -3;
-            playerPos.y = 2;
-            other.gameObject.transform.position = playerPos;
-            if(roomName == "Hall")
+
+            if(roomName == "Hall" && this.name == "library_door")
+            {
+                //setting player postion after they go through a door 
+                SceneManager.LoadScene("Library");
```
Comment (cailafinn) [inline project/sherLOCKED/Assets/Scripts/Objects/DoorInteraction.cs]: This is just an appearance thing, could we make sherry invisible (using `SetActive(false)`) while the new scene is being loaded and the player is being moved? Would just avoid the visual effect of Sherry teleporting around the room.

File: project/sherLOCKED/Assets/Scripts/Objects/DoorInteraction.cs [hunk 7]
```diff
@@ -20,22 +24,90 @@ void OnTriggerEnter2D(Collider2D other)
...
+                playerPos = transform.position;
+                playerPos.x = 3;
+                playerPos.y = 2;
+                other.gameObject.transform.position = playerPos;
+            }
+
+
+            // Only destroy the player when loading the new scene if the player already exists
+            if(numOfPlayers == 1) {
+                DontDestroyOnLoad(other.gameObject);
```
Comment (cailafinn) [inline project/sherLOCKED/Assets/Scripts/Objects/DoorInteraction.cs:144]: A few other things may also need adding to the Don'tDestroyOnLoad, namely the InteractionSymbol, and the ProgressCanvas

**Review by cailafinn — CHANGES_REQUESTED** [review]
Looks good and works well, just some notes on a few little bugs

File: project/sherLOCKED/Assets/Scripts/Objects/DoorInteraction.cs (hunk 6 above)
Comment (Alice-Jaffray) [reply project/sherLOCKED/Assets/Scripts/Objects/DoorInteraction.cs]: Good tip!

File: project/sherLOCKED/Assets/Scripts/Objects/DoorInteraction.cs (hunk 7 above)
Comment (Alice-Jaffray) [reply project/sherLOCKED/Assets/Scripts/Objects/DoorInteraction.cs:144]: Only works for root objects, so not destroying the canvas and it works

**Review by cailafinn — CHANGES_REQUESTED** [review]
Works well, but just noticed the original canvas sticks around when going back to the hall. May need to destroy the existing canvas as well?

**Comment by cailafinn** [comment]
Closing PR replicated via #87

**Pull request closed without merging**